In [1]:
import pandas as pd
import numpy as np
import s3fs

# Milestone 2
___

### Setup your EC2 instance
rubric={correctness:20}

**Please see attached screen shot:**


___

### Setup your JupyterHub¶
rubric={correctness:20}

**Please see attached screen shot:**


___

### Setup the server
rubric={correctness:20}


**Please see attached screenshot:** 


____

### Setup your S3 bucket and move data
rubric={correctness:20}


**Please see attached screenshot:**

___

### Wrangle the data in preparation for machine learning

rubric={correctness:20}

**Reading in Observed Data From S3:**

In [2]:
# AWS credentials
creds = {"key": "ASIATXTRGO6LRTJRLL4E",
         "secret": "4ltnZ8zfOzN0kBTNUePvYNwmdm3Uujm4pjX0MwF3",
        "token": "FwoGZXIvYXdzEFcaDGqhR1ZJbBvt7jNbXSLFAZaJglVtg7DbWamUEg9tQJfTkp4C6coTZzkDjcfTM2sw5G2exjFgrmMXzABGKAqgvGovuIEVRf0P1bzx5u6PA2o/U69ATo87nEWNYcvdtjwX0xaS1Urrruny1mlFiu0naCH6mcG+pSDYAsZ2rkL9DIe0kZ4acO1iHxSADW2gIiaWyTrZAHl5SZ+HlBpSYSueXpAIiT4FgNuSCRGECN/wFSO2Nn1ocm05QaePvL+u1u3Y75QgIGgkP+SNoH4MH0LdaYUPptdOKL3KwpIGMi38Svm5qfCN8TQwJ1rWOk6V8zKrRpzZCcxN3EM79vlB0KZGNrCZ5V+B6QOV8/0="}

# reading in observed data csv file from s3 bucket
observed_data = pd.read_csv("s3://mds-s3-margotvore/output/EC-Earth3-Veg-LR_daily_rainfall_NSW.csv", storage_options=creds)

In [3]:
# setting time as the index and addding a model column to observed data
observed_data.set_index('time', inplace=True)
observed_data['model'] = "Observed Data"
observed_data

lat_min    lat_max   lon_min   lon_max  rain (mm/day)  \
time                                                                           
1889-01-01 12:00:00 -35.887649 -34.766162  141.1875  142.3125   4.139137e+00   
1889-01-02 12:00:00 -35.887649 -34.766162  141.1875  142.3125   7.438660e-03   
1889-01-03 12:00:00 -35.887649 -34.766162  141.1875  142.3125  -5.490768e-20   
1889-01-04 12:00:00 -35.887649 -34.766162  141.1875  142.3125   4.148483e-03   
1889-01-05 12:00:00 -35.887649 -34.766162  141.1875  142.3125   4.291534e-04   
...                        ...        ...       ...       ...            ...   
2014-12-27 12:00:00 -30.280211 -29.158723  152.4375  153.5625   5.162060e+01   
2014-12-28 12:00:00 -30.280211 -29.158723  152.4375  153.5625   4.504323e+01   
2014-12-29 12:00:00 -30.280211 -29.158723  152.4375  153.5625   1.230073e+01   
2014-12-30 12:00:00 -30.280211 -29.158723  152.4375  153.5625   1.943111e-01   
2014-12-31 12:00:00 -30.280211 -29.158723  152.4375  153.5625   7.665157e-01   

                             model  
time                                
1889-01-01 12:00:00  Observed Data  
1889-01-02 12:00:00  Observed Data  
1889-01-03 12:00:00  Observed Data  
1889-01-04 12:00:00  Observed Data  
1889-01-05 12:00:00  Observed Data  
...                            ...  
2014-12-27 12:00:00  Observed Data  
2014-12-28 12:00:00  Observed Data  
2014-12-29 12:00:00  Observed Data  
2014-12-30 12:00:00  Observed Data  
2014-12-31 12:00:00  Observed Data  

[3037320 rows x 6 columns]

**Reading in Model Data From S3:**

In [4]:
# reading in model data parquet file from s3 bucket and setting time as index
model_data = pd.read_parquet("s3://mds-s3-margotvore/combined_models.parquet/", storage_options=creds)
model_data.set_index('time', inplace=True)
model_data = model_data[model_data['model'] != 'observed'] #  removing redundant observed data entries
model_data

lat_min   lat_max  lon_min  lon_max  rain (mm/day)  \
time                                                                        
1889-01-01 12:00:00 -36.250000 -35.00000  140.625  142.500   3.293256e-13   
1889-01-02 12:00:00 -36.250000 -35.00000  140.625  142.500   0.000000e+00   
1889-01-03 12:00:00 -36.250000 -35.00000  140.625  142.500   0.000000e+00   
1889-01-04 12:00:00 -36.250000 -35.00000  140.625  142.500   0.000000e+00   
1889-01-05 12:00:00 -36.250000 -35.00000  140.625  142.500   1.047658e-02   
...                        ...       ...      ...      ...            ...   
2014-12-27 12:00:00 -30.157068 -29.21466  153.125  154.375   5.543748e-01   
2014-12-28 12:00:00 -30.157068 -29.21466  153.125  154.375   7.028577e+00   
2014-12-29 12:00:00 -30.157068 -29.21466  153.125  154.375   2.347570e-01   
2014-12-30 12:00:00 -30.157068 -29.21466  153.125  154.375   2.097459e+00   
2014-12-31 12:00:00 -30.157068 -29.21466  153.125  154.375   5.484206e-01   

                          model  
time                             
1889-01-01 12:00:00  ACCESS-CM2  
1889-01-02 12:00:00  ACCESS-CM2  
1889-01-03 12:00:00  ACCESS-CM2  
1889-01-04 12:00:00  ACCESS-CM2  
1889-01-05 12:00:00  ACCESS-CM2  
...                         ...  
2014-12-27 12:00:00     TaiESM1  
2014-12-28 12:00:00     TaiESM1  
2014-12-29 12:00:00     TaiESM1  
2014-12-30 12:00:00     TaiESM1  
2014-12-31 12:00:00     TaiESM1  

[62467843 rows x 6 columns]

**Concatenating, Filtering, Resampling, and Pivoting Observed and Model Data:**

In [5]:
# Merging the observed data and model data together
filtered_df = pd.concat([model_data, observed_data])

In [6]:
filtered_df

lat_min    lat_max   lon_min   lon_max  rain (mm/day)  \
time                                                                           
1889-01-01 12:00:00 -36.250000 -35.000000  140.6250  142.5000   3.293256e-13   
1889-01-02 12:00:00 -36.250000 -35.000000  140.6250  142.5000   0.000000e+00   
1889-01-03 12:00:00 -36.250000 -35.000000  140.6250  142.5000   0.000000e+00   
1889-01-04 12:00:00 -36.250000 -35.000000  140.6250  142.5000   0.000000e+00   
1889-01-05 12:00:00 -36.250000 -35.000000  140.6250  142.5000   1.047658e-02   
...                        ...        ...       ...       ...            ...   
2014-12-27 12:00:00 -30.280211 -29.158723  152.4375  153.5625   5.162060e+01   
2014-12-28 12:00:00 -30.280211 -29.158723  152.4375  153.5625   4.504323e+01   
2014-12-29 12:00:00 -30.280211 -29.158723  152.4375  153.5625   1.230073e+01   
2014-12-30 12:00:00 -30.280211 -29.158723  152.4375  153.5625   1.943111e-01   
2014-12-31 12:00:00 -30.280211 -29.158723  152.4375  153.5625   7.665157e-01   

                             model  
time                                
1889-01-01 12:00:00     ACCESS-CM2  
1889-01-02 12:00:00     ACCESS-CM2  
1889-01-03 12:00:00     ACCESS-CM2  
1889-01-04 12:00:00     ACCESS-CM2  
1889-01-05 12:00:00     ACCESS-CM2  
...                            ...  
2014-12-27 12:00:00  Observed Data  
2014-12-28 12:00:00  Observed Data  
2014-12-29 12:00:00  Observed Data  
2014-12-30 12:00:00  Observed Data  
2014-12-31 12:00:00  Observed Data  

[65505163 rows x 6 columns]

In [7]:
# Filter data to only include Sydney, Australia, resampling, and pivoting dataframe
filtered_df = filtered_df[filtered_df['lat_min'] <= -33.86]
filtered_df = filtered_df[filtered_df['lat_max'] >= -33.86]
filtered_df = filtered_df[filtered_df['lon_min'] <= 151.21]
filtered_df = filtered_df[filtered_df['lon_max'] >= 151.21]
filtered_df.drop(columns=['lat_min', 'lat_max', 'lon_min', 'lon_max'], inplace=True)
filtered_df.index = pd.to_datetime(filtered_df.index)
filtered_df = filtered_df.groupby('model').resample('1D').mean().reset_index()
filtered_df = filtered_df.pivot(index='time', columns='model', values='rain (mm/day)')
filtered_df

model       ACCESS-CM2  ACCESS-ESM1-5  AWI-ESM-1-1-LR   BCC-CSM2-MR  \
time                                                                  
1889-01-01    0.040427       1.814552    3.557934e+01  4.268112e+00   
1889-01-02    0.073777       0.303965    4.596520e+00  1.190141e+00   
1889-01-03    0.232656       0.019976    5.927467e+00  1.003845e-09   
1889-01-04    0.911319      13.623777    8.029624e+00  8.225225e-02   
1889-01-05    0.698013       0.021048    2.132686e+00  2.496841e+00   
...                ...            ...             ...           ...   
2014-12-27    0.033748       0.123476    1.451179e+00  3.852845e+01   
2014-12-28    0.094198       2.645496    4.249335e+01  5.833801e-01   
2014-12-29    0.005964       3.041667    2.898325e+00  9.359547e-02   
2014-12-30    0.000028       1.131412    2.516381e-01  1.715028e-01   
2014-12-31    0.532747       2.370896    1.047835e-13  4.437736e+00   

model           BCC-ESM1  CMCC-CM2-HR4  CMCC-CM2-SR5  CMCC-ESM2    CanESM5  \
time                                                                         
1889-01-01  1.107466e-03  1.141054e+01  3.322009e-08   2.668800   1.321215   
1889-01-02  1.015323e-04  4.014984e+00  1.312700e+00   0.946211   2.788724   
1889-01-03  1.760345e-05  9.660565e+00  9.103720e+00   0.431999   0.003672   
1889-01-04  1.808932e-01  3.951528e+00  1.317160e+01   0.368693   0.013578   
1889-01-05  4.708019e-09  2.766362e+00  1.822940e+01   0.339267   0.002468   
...                  ...           ...           ...        ...        ...   
2014-12-27  2.061717e-03  8.179260e-09  1.171263e-02   0.090786  59.895053   
2014-12-28  5.939502e-09  8.146937e-01  4.938899e-01   0.000000   0.512632   
2014-12-29  2.000051e-08  2.532205e-01  1.306046e+00   0.000002  37.169669   
2014-12-30  7.191735e-05  8.169252e-02  1.722262e-01   0.788577   7.361246   
2014-12-31  2.863683e-01  6.343592e+00  6.368303e-01   0.442130   0.306608   

model       EC-Earth3-Veg-LR  ...  MPI-ESM-1-2-HAM  MPI-ESM1-2-HR  \
time                          ...                                   
1889-01-01          1.515293  ...     4.244226e-13   1.390174e-13   
1889-01-02          4.771375  ...     4.409552e+00   1.222283e-01   
1889-01-03          4.233980  ...     2.269300e-01   3.762301e-01   
1889-01-04         15.252495  ...     2.344586e-02   4.214019e-01   
1889-01-05         11.920356  ...     4.270161e-13   1.879692e-01   
...                      ...  ...              ...            ...   
2014-12-27          5.071783  ...     4.726998e-13   1.326889e-01   
2014-12-28          1.578188  ...     4.609420e-13   1.644482e+00   
2014-12-29          1.565885  ...     2.016156e+01   1.506439e+00   
2014-12-30          0.025749  ...     9.420543e+00   6.242895e+00   
2014-12-31          0.106525  ...     1.031899e+01   4.765813e+01   

model       MPI-ESM1-2-LR    MRI-ESM2-0         NESM3    NorESM2-LM  \
time                                                                  
1889-01-01   6.537884e-05  3.445495e-06  1.576096e+01  4.759651e-05   
1889-01-02   1.049131e-13  4.791993e-09  3.675510e-01  4.350863e-01   
1889-01-03   9.758706e-14  6.912302e-01  1.562869e-01  9.561101e+00   
1889-01-04   7.060915e-03  3.835721e-02  2.472226e-07  5.301038e-01   
1889-01-05   4.504985e+00  3.506923e-07  1.949792e-13  1.460928e-10   
...                   ...           ...           ...           ...   
2014-12-27   1.827857e+00  6.912632e-03  2.171327e-03  1.620489e+00   
2014-12-28   7.242920e-01  2.836752e-03  1.344768e+01  2.391159e+00   
2014-12-29   1.049481e-01  8.137182e+00  2.547820e+01  1.987695e-12   
2014-12-30   1.245115e-01  9.305263e-03  4.192948e+00  2.150346e+00   
2014-12-31   3.274323e-01  6.854985e-11  2.067847e+00  2.349716e+01   

model       NorESM2-MM  Observed Data  SAM0-UNICON    TaiESM1  
time                                                           
1889-01-01    2.451075       1.515293     0.221324   2.257933  
1889-01-02    0.477231       4.771375     3.757

**Changing Column Order of Output DataFrame:**

In [8]:
# set 'Observed Data' as last column
col_order = [c for c in filtered_df.columns if c not in ["Observed Data"]] + ["Observed Data"]
filtered_df = filtered_df[col_order]
filtered_df

model       ACCESS-CM2  ACCESS-ESM1-5  AWI-ESM-1-1-LR   BCC-CSM2-MR  \
time                                                                  
1889-01-01    0.040427       1.814552    3.557934e+01  4.268112e+00   
1889-01-02    0.073777       0.303965    4.596520e+00  1.190141e+00   
1889-01-03    0.232656       0.019976    5.927467e+00  1.003845e-09   
1889-01-04    0.911319      13.623777    8.029624e+00  8.225225e-02   
1889-01-05    0.698013       0.021048    2.132686e+00  2.496841e+00   
...                ...            ...             ...           ...   
2014-12-27    0.033748       0.123476    1.451179e+00  3.852845e+01   
2014-12-28    0.094198       2.645496    4.249335e+01  5.833801e-01   
2014-12-29    0.005964       3.041667    2.898325e+00  9.359547e-02   
2014-12-30    0.000028       1.131412    2.516381e-01  1.715028e-01   
2014-12-31    0.532747       2.370896    1.047835e-13  4.437736e+00   

model           BCC-ESM1  CMCC-CM2-HR4  CMCC-CM2-SR5  CMCC-ESM2    CanESM5  \
time                                                                         
1889-01-01  1.107466e-03  1.141054e+01  3.322009e-08   2.668800   1.321215   
1889-01-02  1.015323e-04  4.014984e+00  1.312700e+00   0.946211   2.788724   
1889-01-03  1.760345e-05  9.660565e+00  9.103720e+00   0.431999   0.003672   
1889-01-04  1.808932e-01  3.951528e+00  1.317160e+01   0.368693   0.013578   
1889-01-05  4.708019e-09  2.766362e+00  1.822940e+01   0.339267   0.002468   
...                  ...           ...           ...        ...        ...   
2014-12-27  2.061717e-03  8.179260e-09  1.171263e-02   0.090786  59.895053   
2014-12-28  5.939502e-09  8.146937e-01  4.938899e-01   0.000000   0.512632   
2014-12-29  2.000051e-08  2.532205e-01  1.306046e+00   0.000002  37.169669   
2014-12-30  7.191735e-05  8.169252e-02  1.722262e-01   0.788577   7.361246   
2014-12-31  2.863683e-01  6.343592e+00  6.368303e-01   0.442130   0.306608   

model       EC-Earth3-Veg-LR  ...  MPI-ESM-1-2-HAM  MPI-ESM1-2-HR  \
time                          ...                                   
1889-01-01          1.515293  ...     4.244226e-13   1.390174e-13   
1889-01-02          4.771375  ...     4.409552e+00   1.222283e-01   
1889-01-03          4.233980  ...     2.269300e-01   3.762301e-01   
1889-01-04         15.252495  ...     2.344586e-02   4.214019e-01   
1889-01-05         11.920356  ...     4.270161e-13   1.879692e-01   
...                      ...  ...              ...            ...   
2014-12-27          5.071783  ...     4.726998e-13   1.326889e-01   
2014-12-28          1.578188  ...     4.609420e-13   1.644482e+00   
2014-12-29          1.565885  ...     2.016156e+01   1.506439e+00   
2014-12-30          0.025749  ...     9.420543e+00   6.242895e+00   
2014-12-31          0.106525  ...     1.031899e+01   4.765813e+01   

model       MPI-ESM1-2-LR    MRI-ESM2-0         NESM3    NorESM2-LM  \
time                                                                  
1889-01-01   6.537884e-05  3.445495e-06  1.576096e+01  4.759651e-05   
1889-01-02   1.049131e-13  4.791993e-09  3.675510e-01  4.350863e-01   
1889-01-03   9.758706e-14  6.912302e-01  1.562869e-01  9.561101e+00   
1889-01-04   7.060915e-03  3.835721e-02  2.472226e-07  5.301038e-01   
1889-01-05   4.504985e+00  3.506923e-07  1.949792e-13  1.460928e-10   
...                   ...           ...           ...           ...   
2014-12-27   1.827857e+00  6.912632e-03  2.171327e-03  1.620489e+00   
2014-12-28   7.242920e-01  2.836752e-03  1.344768e+01  2.391159e+00   
2014-12-29   1.049481e-01  8.137182e+00  2.547820e+01  1.987695e-12   
2014-12-30   1.245115e-01  9.305263e-03  4.192948e+00  2.150346e+00   
2014-12-31   3.274323e-01  6.854985e-11  2.067847e+00  2.349716e+01   

model       NorESM2-MM  SAM0-UNICON    TaiESM1  Observed Data  
time                                                           
1889-01-01    2.451075     0.221324   2.257933       1.515293  
1889-01-02    0.477231     3.757179   2.287381 

In [9]:
filtered_df.shape

(46020, 26)

**The shape of the ML-ready data is** `(46020, 26)`

In [10]:
filtered_df.to_csv(
    "s3://mds-s3-margotvore/output/ml_data.csv", storage_options=creds
)